### DISCUSSION TITLE
purpose of _proto_info cache in grpc-gateway/deco.lua

### DISCUSSION THREAD

**Discussion by git-torrent** [Kong/kong#8421] [Help]
In file https://github.com/Kong/kong/blob/master/kong/plugins/grpc-gateway/deco.lua#L81 file is recursively parsed by protoc and features are extracted.

@javierguerragiraldez What is the reason behind this solution? Instead of parsing `p.loaded` there is certain form of cache filled up with all the methods of requested file and all its imports. 

Isn't it faster to parse `p.loaded` while looking just for a `method` actually being called?

Thanks for explanation

**Comment by javierguerragiraldez** [comment]
currently, protoc doesn't present the services and methods of imported proto files.  only message and enum definitions are presented in the returned object.

also, this isn't looking for the method actually being called, it caches every method defined so it's done only once.

**Reply by git-torrent** [reply]
Thanks for answer @javierguerragiraldez 

I think this originates from times before `include_imports` directive was added (commit 70d35871d1d51205601ba99ce871eb134b06e395, namely https://github.com/Kong/kong/blob/70d35871d1d51205601ba99ce871eb134b06e395/kong/plugins/grpc-gateway/deco.lua#L94). In the current situation the files are in the `p.loaded` table accessible by `p.loaded[ file.dependency[ i+1 ] ]` (`i` goes through `public_dependency` list). Is the `parsefile` step still necessary?

I also wonder why the function grpc(_tools).each method https://github.com/Kong/kong/blob/master/kong/tools/grpc.lua#L56 has param recurse. Can it be false in some instances?
